### PR TITLE
Merge DOI services

### PR DESCRIPTION
**Pull request by janbj**
This is a huge pull request and the work getting here was not trivial. the NVA DOI Transformer did not have code coverage requirements and I needed to take some shortcuts in the end. This should be improved as we continue the work.

### CODE REVIEW

File: doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param contentType contentType
+     * @return DataciteContentType
+     */
+    public static DataciteContentType lookup(String contentType) {
+        return Arrays
+            .stream(DataciteContentType.values())
+            .filter(dataciteContentType -> dataciteContentType.getContentType().equals(contentType))
+            .findAny()
+            .orElseThrow(() ->
```
Comment (brinxmat) [inline doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java:35]: This entire thing is challenging to understand, extract a constant: 
`    private static final String ALL_CONTENT_TYPES = Arrays.stream(values())
            .map(DataciteContentType::getContentType)
            .collect(Collectors.joining(", "));`

Comment (brinxmat) [reply doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java:35]: We should probably log these errors

File: doi-proxy/src/main/java/no/unit/nva/doi/DoiProxyService.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @param dataciteContentType   dataciteContentType
+     * @return  metadata and content location
+     * @throws MetadataNotFoundException    when no metadata found
+     * @throws IOException  when error reading from IO
+     * @throws URISyntaxException   when malformed doi url
+     */
+    public MetadataAndContentLocation lookupDoiMetadata(String doiUrl, DataciteContentType dataciteContentType)
+        throws MetadataNotFoundException, IOException, URISyntaxException {
+        logger.info("getDoiMetadata(doi:" + doiUrl + ")");
+        MetadataAndContentLocation metadataAndContentLocation = null;
```
Comment (brinxmat) [inline doi-proxy/src/main/java/no/unit/nva/doi/DoiProxyService.java:53]: Redundant initializer

File: doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java (hunk 1 above)
Comment (brinxmat) [reply doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java:35]: Also SingletonCollector?

File: doi-proxy/src/test/java/no/unit/nva/doi/model/DoiProxyResponseTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import no.unit.nva.doi.fetch.ObjectMapperConfig;
+import no.unit.nva.doi.fetch.model.utils.MetadataSource;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class DoiProxyResponseTest {
+
+    private ObjectMapper objectMapper = ObjectMapperConfig.createObjectMapper();
+
+    @Test
```
Comment (brinxmat) [inline doi-proxy/src/test/java/no/unit/nva/doi/model/DoiProxyResponseTest.java:14]: Is this test more "assertDoesNotThrow"?

File: doi-proxy/src/test/java/no/unit/nva/doi/utils/AbstractLambdaTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+public abstract class AbstractLambdaTest {
+
+    public static final String DOI_STRING = "10.1007/s00115-004-1822-4";
+    public static final String DOI_DX_URL_PREFIX = "https://dx.doi.org";
+    public static final String DOI_URL_PREFIX = "https://doi.org";
+
+    public static final Path CrossRefSamplePath = Paths.get("crossRefSample.json");
+    public static final String ERROR_MESSAGE = "404 error message";
+
+    protected final Context mockLambdaContext = createMockContext();
```
Comment (brinxmat) [inline doi-proxy/src/test/java/no/unit/nva/doi/utils/AbstractLambdaTest.java:25]: Unused

File: doi-proxy/src/test/java/no/unit/nva/doi/utils/MockHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    public MockHttpClient(AbstractHttpResponse<R> response) {
+        this.response = response;
+    }
+
+    // T must be the same with R
+    @Override
+    public <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest request,
+                                                            BodyHandler<T> responseBodyHandler) {
+
+        CompletableFuture<HttpResponse<T>> result = CompletableFuture.completedFuture(
```
Comment (brinxmat) [inline doi-proxy/src/test/java/no/unit/nva/doi/utils/MockHttpClient.java:31]: return inline

Comment (brinxmat) [reply doi-proxy/src/test/java/no/unit/nva/doi/utils/MockHttpClient.java:31]: Annotate for unchecked

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    private int parseSequence(String sequence, int alternativeSequence) {
+        RuleBasedNumberFormat nf = new RuleBasedNumberFormat(Locale.UK, RuleBasedNumberFormat.SPELLOUT);
+        try {
+            return nf.parse(sequence).intValue();
+        } catch (Exception e) {
+            return alternativeSequence;
+        }
+    }
+
+    private List<String> extractTags() {
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java:314]: Set up a task to do something with the methods that return null in CrossrefConverter

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    protected boolean hasOpenAccessRights(DataciteRights dataciteRights) {
+        return Optional.ofNullable(dataciteRights.getRightsUri())
+            .map(LicensingIndicator::isOpen).orElse(false);
+    }
+
+    private PublicationType extractPublicationType(DataciteResponse dataciteResponse) {
+        return DataciteTypesUtil.mapToType(dataciteResponse);
+    }
+
+    private String extractAbstract() {
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/DataciteResponseConverter.java:230]: Set up a task to do something with the methods that return null in DataciteConverter

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/DoiTransformService.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+     * @param orgNumber       the orgNumber.
+     * @return a Publication.
+     * @throws JsonProcessingException  when cannot process json.
+     * @throws MissingClaimException     when request does not have the required claims.
+     * @throws URISyntaxException       when the input contains invalid URIs
+     * @throws InvalidIssnException     thrown if a provided ISSN is invalid.
+     * @throws InvalidPageTypeException thrown if the provided page type is incompatible with
+     *                                  the publication instance type.
+     */
+    public Publication transformPublication(String body, String contentLocation, String owner, String orgNumber)
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/DoiTransformService.java:60]: Remove unthrown exception

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/MetadataLocation.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+public enum MetadataLocation {
+    CROSSREF("https://api.crossref.org"), DATACITE("https://data.datacite.org");
+
+    public static final String CROSSREF_STRING = "crossref";
+    public static final String DATACITE_STRING = "datacite";
+    private static String ILLEGAL_ARGUMENT_ERROR = "Illegal MetadataLocation value. Valid values are:";
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/MetadataLocation.java:12]: final

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/MetadataLocation.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    private static Map<String, MetadataLocation> valuesMap;
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/MetadataLocation.java:13]: final

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/language/LanguageMapper.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        String pathString = path.toString();
+        return Thread.currentThread().getContextClassLoader().getResourceAsStream(pathString);
+    }
+
+    private static InputStreamReader newInputStreamReader(InputStream stream) {
+        return new InputStreamReader(stream, StandardCharsets.UTF_8);
+    }
+
+    @SuppressWarnings("PMD.AvoidProtectedMethodInFinalClassNotExtending")
+    protected static List<String> linesfromResource(Path path) {
```
Comment (brinxmat) [inline doi-transformer/src/main/java/no/unit/nva/doi/transformer/language/LanguageMapper.java:100]: Probably from commons…again

File: doi-transformer/src/test/java/no/unit/nva/doi/transformer/utils/TestLambdaLogger.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package no.unit.nva.doi.transformer.utils;
```
Comment (brinxmat) [inline doi-transformer/src/test/java/no/unit/nva/doi/transformer/utils/TestLambdaLogger.java:1]: Commons supports this

File: build.gradle [hunk 13]
```diff
@@ -1,86 +1,114 @@
...
-
 group 'no.unit.nva'
 version '1.0-SNAPSHOT'
 
+allprojects {
+    apply plugin: 'nebula.lint'
 
-sourceCompatibility = JavaVersion.VERSION_11
-targetCompatibility = JavaVersion.VERSION_11
+    gradleLint.rules = ['unused-dependency']
```
Comment (axthosarouris) [inline build.gradle:15]: why not all-dependency?

File: build.gradle [hunk 14]
```diff
@@ -1,86 +1,114 @@
...
-
-test {
-    useJUnitPlatform()
-    testLogging {
-        events "passed", "skipped", "failed"
+    test {
+        useJUnitPlatform()
+        failFast = true
+        testLogging {
+            events 'skipped', 'passed', 'failed'
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply build.gradle:61]: You can put parenthesis, square brackets do not compile

File: .gitignore [hunk 15]
```diff
@@ -281,4 +281,6 @@ gradle-app.setting
 ### Gradle Patch ###
 **/build/
 
-# End of https://www.gitignore.io/api/java,linux,macos,gradle,windows,eclipse,intellij+all,visualstudiocode
\ No newline at end of file
+# End of https://www.gitignore.io/api/java,linux,macos,gradle,windows,eclipse,intellij+all,visualstudiocode
+/packaged.yaml
```
Comment (axthosarouris) [inline .gitignore:285]: Can you put a comment what are those files?

File: build.gradle [hunk 16]
```diff
@@ -1,86 +1,114 @@
...
+
+    sourceCompatibility = JavaVersion.VERSION_11
+    // source-code version and must be <= targetCompatibility
+    targetCompatibility = JavaVersion.VERSION_11 // bytecode target version
+
+    def junit5Version = '5.6.0'
+
+    dependencies {
+        testImplementation group: 'com.github.BIBSYSDEV', name: 'nva-testutils', version: 'v0.1.9'
+        testImplementation(group: 'junit', name: 'junit', version: '4.13') {
```
Comment (axthosarouris) [inline build.gradle:37]: I have removed all dependencies from old junit. see my PR.

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java [hunk 17]
```diff
@@ -53,7 +55,7 @@ public PublicationResponse insertPublication(Publication publication, String api
         //TODO: toResponse is exactly the same method as convertValue and will be changed upon next release
```
Comment (axthosarouris) [inline create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java:55]: I would recommend that we create a task and we put the task number in the TODO

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+import nva.commons.handlers.RequestInfo;
+import nva.commons.utils.Environment;
+import nva.commons.utils.RequestUtils;
+import org.apache.http.HttpHeaders;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MainHandler extends ApiGatewayHandler<RequestBody,Summary> {
+
+    public static final String API_HOST_ENV = "API_HOST";
```
Comment (axthosarouris) [inline create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java:40]: Can we rename this to indicate the role of the   micro service we are calling?

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        this.doiProxyService = doiProxyService;
+        this.publicationPersistenceService = publicationPersistenceService;
+        this.apiHost = environment.readEnv(API_HOST_ENV);
+        this.apiScheme = environment.readEnv(API_SCHEME_ENV);
+    }
+
+    @Override
+    protected Summary processInput(RequestBody input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+        String apiUrl = String.join("://", apiScheme, apiHost);
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java:85]: We could use a URI builder here and build a URI and not a String. I recommend to create an issue (bug?) for this

File: create-publication-from-doi/src/test/java/no/unit/nva/doi/fetch/MainHandlerTest.java [hunk 20]
```diff
@@ -126,12 +135,11 @@ private PublicationConverter mockPublicationConverter() {
     }
 
     private DoiTransformService mockDoiTransformServiceReturningSuccessfulResult()
-        throws URISyntaxException, TransformFailedException, InterruptedException, IOException,
+        throws URISyntaxException, IOException,
```
Comment (axthosarouris) [inline create-publication-from-doi/src/test/java/no/unit/nva/doi/fetch/MainHandlerTest.java:138]: push all private methods that are not tests to the bottom

File: create-publication-from-doi/src/test/java/no/unit/nva/doi/fetch/MainHandlerTest.java [hunk 21]
```diff
@@ -148,13 +156,17 @@ private Publication getPublication() {
...
-        throws NoContentLocationFoundException, InterruptedException, IOException, URISyntaxException,
-        MetadataNotFoundException, MalformedRequestException {
+        throws MetadataNotFoundException, IOException, URISyntaxException {
         DoiProxyService doiProxyService = mock(DoiProxyService.class);
-        when(doiProxyService.lookup(any(), anyString(), anyString())).thenReturn(mockDoiProxyResponse());
+        when(doiProxyService.lookupDoiMetadata(anyString(), any())).thenReturn(metadataAndContentLocation());
         return doiProxyService;
     }
 
+    private MetadataAndContentLocation metadataAndContentLocation() throws JsonProcessingException {
```
Comment (axthosarouris) [inline create-publication-from-doi/src/test/java/no/unit/nva/doi/fetch/MainHandlerTest.java:165]: Line 164: Remove unused method `mockDoiProxyResponse()`

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+    protected Integer getSuccessStatusCode(RequestBody input, Summary output) {
+        return SC_OK;
+    }
+
+    private Publication getPublicationMetadata(RequestBody requestBody,
+                                            String owner, String orgNumber)
+        throws URISyntaxException, IOException, MissingClaimException, InvalidIssnException,
+               InvalidPageTypeException, MetadataNotFoundException {
+        MetadataAndContentLocation metadataAndContentLocation = doiProxyService.lookupDoiMetadata(
+            requestBody.getDoiUrl().toString(), DataciteContentType.DATACITE_JSON);
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/MainHandler.java:131]: This is a header for the expected format (JSON or XML) response for Datacite only. Crossref returns only JSON.

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java [hunk 23]
```diff
@@ -26,13 +28,14 @@
     public static final String WARNING_MESSAGE = "Inserting publication failed.";
     public static final String INSERTING_PUBLICATION_FAILED = WARNING_MESSAGE + "\nAPI-URL:%s\nRequestBody:%s\n";
     private final HttpClient client;
-    private final ObjectMapper objectMapper = MainHandler.createObjectMapper();
+    private final ObjectMapper objectMapper = ObjectMapperConfig.createObjectMapper();
```
Comment (axthosarouris) [inline create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java:31]: Do we need this objectMapper here? Can we use the nva commons mapper? I remember we had an issue that the JsonUtils mapper is too strict for when we are importing data from outside. Is this the reason why we use this one here?

File: doi-proxy/src/main/java/no/unit/nva/doi/DataciteClient.java [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+                                   doiUrl.getPath()));
+    }
+
+    public MetadataAndContentLocation fetchMetadata(String doiUrlString, DataciteContentType dataciteContentType)
+        throws IOException {
+        String json = readStringFromUrl(createRequestUrl(doiUrlString, dataciteContentType));
+        return new MetadataAndContentLocation(DATACITE_BASE_URL_STRING, json);
+    }
+
+    protected String readStringFromUrl(URL url) throws IOException {
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply doi-proxy/src/main/java/no/unit/nva/doi/DataciteClient.java:34]: https://unit.atlassian.net/browse/NP-1112

File: doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java (hunk 1 above)
Comment (axthosarouris) [reply doi-proxy/src/main/java/no/unit/nva/doi/DataciteContentType.java:35]: I think we should add a method to SingletonCollector `orElseThrow()`

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java (hunk 23 above)
Comment (janbj) [reply create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java:31]: Yes, we can use the one from nva-commons. I started refactoring without using nva-commons (tried to postpone to other PR) and then I ended up with having to use it.

File: create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java (hunk 17 above)
Comment (janbj) [reply create-publication-from-doi/src/main/java/no/unit/nva/doi/fetch/service/PublicationPersistenceService.java:56]: It is resolved now, removing comment :)

File: build.gradle (hunk 13 above)
Comment (janbj) [reply build.gradle:16]: Rurik recommened this setting.

File: doi-transformer/src/main/java/no/unit/nva/doi/transformer/language/LanguageMapper.java (hunk 11 above)
Comment (janbj) [reply doi-transformer/src/main/java/no/unit/nva/doi/transformer/language/LanguageMapper.java:100]: Already in commons or move to commons?

File: fetch-doi-commons/build.gradle [hunk 25]
```diff
@@ -17,7 +17,7 @@ dependencies {
         because 'Used by org.apache.logging.log4j:log4j-slf4j18-impl:2.13.2 in nva-commons'
     }
     api 'com.github.BIBSYSDEV:aws-build-tools:v1.5.5'
-    api group: 'com.github.BIBSYSDEV', name: 'nva-datamodel-java', version: 'v0.8.1'
+    api group: 'com.github.BIBSYSDEV', name: 'nva-datamodel-java', version: 'v0.9.0'
```
Comment (brinxmat) [inline fetch-doi-commons/build.gradle:20]: Do we still use "v"?

**Pull request merged**